File: 53.maximum-subarray/solution.cpp
class Solution {
public:
	int maxSubArray(vector<int>& nums) {
		int len = nums.size();
		int maxresult = nums.front();
		int val = maxresult;
		bool inited = false;
		int max = val;
		int min = val;
		for (int i = 1; i < len; i++)
		{
			int &v = nums[i];
			inited = true;
			val += v;
			if (val < min)
			{
				min = val;
				max = val;
				inited = false;
				maxresult = v > maxresult ? v : maxresult;
			} else
				max = val > max ? val : max;
			int result = inited ? max - (min < 0 ? min : 0) : max;
			maxresult = result > maxresult ? result : maxresult;
		}
		return maxresult;
	}
};

File: 53.maximum-subarray/solution.py
class Solution:
    def maxSubArray(self, nums: List[int]) -> int:
        sums = [0] * len(nums)
        sums[0] = nums[0]
        for i in range(1, len(nums)):
            if sums[i - 1] < 0:
                sums[i] = nums[i]
            else:
                sums[i] = sums[i - 1] + nums[i]
        return max(sums)
